aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/S3StorageTransferClientProvider.kt

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.storage.s3.transfer
+
+import aws.sdk.kotlin.services.s3.S3Client
+import com.amplifyframework.auth.AuthCredentialsProvider
+import com.amplifyframework.storage.StorageException
+
+internal class S3StorageTransferClientProvider(
+    private val createS3Client: (region: String?, bucketName: String?) -> S3Client
+) : StorageTransferClientProvider {
+    override fun getStorageTransferClient(region: String?, bucketName: String?): S3Client {
```

Comment:
Do we need to also inject the credentials provider as that is a global s3 config option I believe.

Reply:
Another thought. At this point, could we make region and bucketName required? We should be able to provide the default values for both of these at any point that this method is getting called. It would allow us to remove all of the fallback logic elsewhere if region and bucketName were no longer null here.

Reply:
The clientProvider is mainly/directly used to create S3 clients from the transfer records which doesn't have context or access to to credentials. 
The transfer record has region and bucket name fields as null by default, so bucket name and region can't be required here as it is possible for either of those fields to be null when reading from the database.
Here's an example of usage:
https://github.com/aws-amplify/amplify-android/pull/2899/files#diff-9fdcc6ff34286fc160d66a1b782c065eb10867f6c21e70e2b37fdf9287ba3f17R44

Reply:
I don't think we currently expose it publicly, its mainly for tests. But the AWSS3StoragePlugin takes a global credentialsProvider. We inject that provider into the s3 client. The S3Client from within the Kotlin SDK will then resolve the credentials. All of these created S3Clients are missing a way to inject the proper credentials provider provided in the plugin constructor.

```
private var s3Client: S3Client = S3Client {
        region = awsRegion
        credentialsProvider = authCredentialsProvider
}
```

Reply:
Discussed offline and verified that all `S3Client` instances have the appropriate region and credentials injected.